tile_pages.py

```diff
@@ -255,79 +275,135 @@ def run(self,rows=None,cols=None):
                 continue
 
             if self.col_major:
-                c = math.floor(i/rows)
-                r = i % rows
+                c = math.floor(i/self.rows)
+                r = i % self.rows
             else:
-                r = math.floor(i/cols)
-                c = i % cols
+                r = math.floor(i/self.cols)
+                c = i % self.cols
             
             if self.right_to_left:
-                c = cols - c - 1
+                c = self.cols - c - 1
+            
+            if self.bottom_to_top:
+                r = self.rows - r - 1
+                        
+            scale_factor = 1
+            scale_shift_right = 0
+            scale_shift_up = 0
             
-            if not self.bottom_to_top:
-                r = rows - r - 1
+            if page_box_defined:
+                # calculate scaling factors based on source page size
+                source_width = pw[i]
+                source_height = ph[i]
+                scalef_width = page_box_width / source_width
+                scalef_height = page_box_height / source_height
+                # take the smaller scaling factor so that the page will fit into its box
+                if scalef_width <= scalef_height:
+                    scale_factor = scalef_width
+                else:
+                    scale_factor = scalef_height
             
-            x0 = margin - trim[0] + c*(pw - trim[0] - trim[1]) + o_shift[0]
-            y0 = margin - trim[3] + r*(ph - trim[2] - trim[3]) + o_shift[1]
-
-            # don't scale, just shift and rotate
+            if page_box_defined:
+                cpos_x0 = c*page_box_width
+                cpos_y0 = (self.rows-r-1)*page_box_height
+            else:
+                cpos_x0 = sum(pw[r*self.cols:r*self.cols + c])
+                cpos_y0 = sum(row_height[r+1:])
+            
+            x0 = margin - trim[0] + cpos_x0 - c*(trim[0] + trim[1])
+            y0 = margin - trim[3] + cpos_y0 - (self.rows-r-1)*(trim[2] + trim[3])
+            
+            if page_box_defined and self.center_content:
+                # center pages in their box
+                scaled_width = pw[i] * scale_factor
+                scaled_height = ph[i] * scale_factor
+                # unless we are using round here, there is no content - for whatever reason
+                shift_right = round((page_box_width-scaled_width)/2)
+                shift_up = round((page_box_height-scaled_height)/2)
+                
+                x0 += shift_right
+                y0 += shift_up
+                        
+            if scale_factor != 1:
+                performed_scale = True
+            
+            # scale, shift and rotate
             # first shift to origin, then rotate, then shift to final destination
-            content_txt += f'q {R[0]} {R[1]} {R[2]} {R[3]} {x0} {y0} cm '
+            content_txt += f'q {scale_factor} {R[1]} {R[2]} {scale_factor} {x0+o_shift[0]} {y0+o_shift[1]} cm '
```

Comment:
I don't think this will work with rotation - for example if rotation is clockwise then R = [0, -1, 1 0], but with scale_factor = 1 this would write [1, -1, 1, 1] instead. I think you need to do the matrix multiplication between the rotation matrix and scaling matrix.

Reply:
Yes, that's a big mistake. Files indeed look completely wrong when setting rotation. I need to change this.

Reply:
Hmm, I attemted to address the problem, but there's still a shift in y direction that I don't know how to compensate for.

Reply:
It should be fixed now I hope.